compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -1,46 +1,55 @@
-//! Note: tests specific to this file can be found in:
+//! # Match exhaustiveness and reachability algorithm
 //!
-//!   - `ui/pattern/usefulness`
-//!   - `ui/or-patterns`
-//!   - `ui/consts/const_in_pattern`
-//!   - `ui/rfc-2008-non-exhaustive`
-//!   - `ui/half-open-range-patterns`
-//!   - probably many others
+//! This file contains the logic for exhaustiveness and reachability checking for pattern-matching.
+//! Specifically, given a list of patterns in a match, we can tell whether:
+//! (a) a given pattern is reachable (reachability)
+//! (b) the patterns cover every possible value for the type (exhaustiveness)
 //!
-//! I (Nadrieril) prefer to put new tests in `ui/pattern/usefulness` unless there's a specific
-//! reason not to, for example if they depend on a particular feature like `or_patterns`.
+//! The algorithm implemented here is inspired from the one described in [this
+//! paper](http://moscova.inria.fr/~maranget/papers/warn/index.html). We have however changed it in
```

Comment:
Perhaps once this change is merged, it would be worth revisiting the idea of writing a paper about the pattern matching algorithm in rustc, given how much has changed from algorithm described in the original paper...

Reply:
I was just thinking that the other day! Specifically because I'm working on never patterns atm which are a very unique pattern-matching feature. But you're right what we already have is pretty rad.

Full honesty though: I'd be more than happy to assist someone, but I'm not writing a paper again if I can help it x)

Reply:
Would you like to write it? I'd have time to help you these days

Reply:
I'd be interested in working on one, but don't have the time at the moment. Perhaps some time in the new year?

Reply:
Let's keep in touch :D